### PR TITLE
Per-space notification bubbles

### PR DESCRIPTION
**Pull request by LorenDB**
- Display unread notifications for spaces
- Fix typoed variable name
- Add space notifications to room list
- Add loud notifications for spaces
- Make the notification bubble its own component
- Add space notifs to room list
- make lint
- Add space notification configuration

### CODE REVIEW

**Comment by LorenDB** [comment]
This does not support notifications for tags and categories; I'm not sure if there is a good way to detect what rooms are under specific tags and categories and I don't know if I really want to look at that right now.

**Comment by LorenDB** [comment]
Okey dokey, I've got tags, DMs, and the like hooked up. However, before merging, I think a review of the configuration for this would be in order. In other words, how much should the user be able to turn off and on?

**Comment by deepbluev7** [comment]
I think ideally we would always show notification counts, but be able to mute on a per space bases, which would hide the bubble (and notifications) for that specific space and all the children?

**Comment by LorenDB** [comment]
Sounds reasonable to me.

**Comment by LorenDB** [comment]
> I think ideally we would always show notification counts, but be able to mute on a per space bases, which would hide the bubble (and notifications) for that specific space and all the children?

Done, but there are some problems with toggling the space mute (see my comment in #nheko:nheko.im).

File: resources/qml/CommunitiesList.qml [hunk 1]
```diff
@@ -36,14 +37,27 @@ Page {
...
+                hidden = hidden_;
+                muted = muted_;
                 open();
             }
 
+            Platform.MenuItem {
+                text: qsTr("Do not show notification counts for this space or tag.")
+                checkable: true
+                checked: communityContextMenu.muted
+                onTriggered: Communities.toggleTagMute(communityContextMenu.tagId)
```
Comment (deepbluev7) [inline resources/qml/CommunitiesList.qml:54]: This doesn't seem to work?

**Comment by deepbluev7** [comment]
Looks good to me apart from the tiny thing, that the menu entry for muting a space doesn't seem to work :3

**Comment by deepbluev7** [comment]
Hm, recalculating the space unread counts seems to take quite a while whenever I am switching rooms...

File: src/timeline/CommunitiesModel.cpp [hunk 2]
```diff
@@ -70,6 +92,17 @@ CommunitiesModel::data(const QModelIndex &index, int role) const
             return 0;
         case CommunitiesModel::Roles::Id:
             return "";
+        case CommunitiesModel::Roles::UnreadMessages: {
+            int total{0};
+            for (const auto &[id, info] : cache::getRoomInfo(cache::joinedRooms()))
```
Comment (deepbluev7) [inline src/timeline/CommunitiesModel.cpp:97]: This is really, really slow. Every time I click on a room, this causes `900*900*2=1620000` database transactions and makes everything slow. There is a command to fetch all rooms at once, but I don't think you should even fetch all rooms here, we are already caching this stuff somewhere.

File: src/timeline/CommunitiesModel.cpp [hunk 3]
```diff
@@ -84,16 +117,27 @@ CommunitiesModel::data(const QModelIndex &index, int role) const
...
+            return hiddenTagIds_.contains(QStringLiteral("dm"));
         case CommunitiesModel::Roles::Parent:
             return "";
         case CommunitiesModel::Roles::Depth:
             return 0;
         case CommunitiesModel::Roles::Id:
             return "dm";
+        case CommunitiesModel::Roles::UnreadMessages: {
+            int total{0};
+            for (const auto &[id, info] : cache::getRoomInfo(directMessages_))
```
Comment (deepbluev7) [inline src/timeline/CommunitiesModel.cpp:129]: Similarly, this can be fetched in one call from the db.

Comment (deepbluev7) [reply src/timeline/CommunitiesModel.cpp:129]: (Or just access the rooms directly)

**Comment by LorenDB** [comment]
> Looks good to me apart from the tiny thing, that the menu entry for muting a space doesn't seem to work :3

I think it is simply a problem with QML not querying the model when updates happen; I tested with the notification visibility being dependent on hiddenness instead of mutedness and it didn't work there.

File: src/timeline/CommunitiesModel.cpp (hunk 3 above)
Comment (LorenDB) [reply src/timeline/CommunitiesModel.cpp:129]: > (Or just access the rooms directly)

Via...?

**Comment by deepbluev7** [comment]
> I think it is simply a problem with QML not querying the model when updates happen; I tested with the notification visibility being dependent on hiddenness instead of mutedness and it didn't work there.

If you specify nothing, it considers everything changed.

**Comment by LorenDB** [comment]
I just fixed that.

I'm personally at a loss as to how to fix this because the menu's show function does pick up the change; however, QML never rerenders the notification bubbles until you restart.

**Comment by LorenDB** [comment]
See also #1090

**Comment by deepbluev7** [comment]
Well, it is not really a blocker for me, the performance is though.

**Comment by deepbluev7** [comment]
(Also you could try debugging using Gammaray or so)

**Comment by LorenDB** [comment]
Thanks for fixing my code!

**Pull request merged**